plugin-server/src/cdp/cdp-api.ts

```diff
@@ -193,4 +201,69 @@ export class CdpApi {
             res.status(500).json({ error: e.message })
         }
     }
+
+    private postFunctionInvocationReal = async (req: express.Request, res: express.Response): Promise<void> => {
+        try {
+            const { id, team_id } = req.params
+            const { invocations }: { invocations: HogFunctionInvocationGlobals[] } = req.body
+
+            status.info('⚡️', 'Received invocations', { id, team_id, body: req.body })
+
+            if (!invocations || !invocations.length) {
+                res.status(400).json({ error: 'Missing invocations' })
+                return
+            }
+
+            const [hogFunction, team] = await Promise.all([
+                this.hogFunctionManager.fetchHogFunction(req.params.id),
+                this.hub.teamManager.fetchTeam(parseInt(team_id)),
+            ]).catch(() => {
+                return [null, null]
+            })
+            if (!hogFunction || !team || hogFunction.team_id !== team.id) {
+                res.status(404).json({ error: 'Hog function not found' })
+                return
+            }
+
+            // Enqueue all the invocations
+            const preparedInvocations = invocations.map((x) =>
+                createInvocation(
+                    {
+                        person: x.person,
+                        event: x.event,
+                        project: {
+                            id: team.id,
+                            name: team.name,
+                            url: `${this.hub.SITE_URL ?? 'http://localhost:8000'}/project/${team.id}`,
+                        },
+                    },
+                    hogFunction,
+                    // The "email" hog functions export a "sendEmail" function that we must explicitly call
+                    // BW - This feels super tricky. We have specific functions for specific things? Why don't we just invoke the function as it is with the required globals like any other function?
```

Comment:
This is something at the v0 level of quality. If we go down this path, we'll definitely need some systems to organize/map which paths should export which functions.

In an earlier version [I had a Hog STL function](https://github.com/PostHog/posthog/pull/25719/files#diff-523d5432cf6383c2555371f83ab04b3fdcd7ac8704994e39fb5b0e2c456848f3) under `hogvm/stl/src/sendEmail.hog`, essentially:

```rust
fun sendEmail(email) {
  let provider := __importCallable('provider/email')() // this was some earlier version of the syntax
  return provider.sendEmail(email)
}
```

but removed it in the end. It felt too tied to PostHog's internals to be part of the core Hog STL. I wanted to have a discussion on what's a good approach here, but for v0 kept as you see it now.

---

When you "import" a path we do execute the code directly, and we also return whatever the top level function returns as the response of `import('file')`. The way we export (returning an object at the end of the code) is quite yuk. I want to add an `export` keyword in front of `let`/`fun` to make it automatic.

Initially I did want to just run the code with some extra global variables (`inputs.args`? `args`?), but in addition to some technical points around merging globals, all of that felt a bit too hard to follow at the end of the day (where does what come from?). While _that_ could still be managed, when thinking a few steps further, I realized users will soon want to encapsulate some code through shared utility files/functions. The ergonomics of that would get rather annoying if we would impose what would amount to a "one exportable function per file" policy. For example I'd like to export both `getApiURL()` and `makeApiRequest()` from my `utils/api.hog` file.

Taking all of that into account, the simplest way to make sharing providers work for the messaging system was to develop a proper code import/export system (minus the `export` keyword for now). The drawback is that we now need to specify in the code that `provider/email` exports `sendEmail` when testing the function.